pkg/protodep/gomod.go

```diff
@@ -0,0 +1,331 @@
+package protodep
+
+import (
+	"bufio"
+	"context"
+	"fmt"
+	"os"
+	"path/filepath"
+	"strings"
+	"unicode"
+
+	zglob "github.com/mattn/go-zglob"
+	"github.com/pkg/errors"
+	"github.com/solo-io/solo-kit/pkg/utils/modutils"
+	"github.com/spf13/afero"
+)
+
+const (
+	ProtoMatchPattern   = "**/*.proto"
+	SoloKitMatchPattern = "**/solo-kit.json"
+)
+
+var (
+	// offer sane defaults for proto vendoring
+	DefaultMatchPatterns = []string{ProtoMatchPattern, SoloKitMatchPattern}
+
+	// matches ext.proto for solo hash gen
+	ExtProtoMatcher = &GoModImport{
+		Package:  "github.com/solo-io/protoc-gen-ext",
+		Patterns: []string{"extproto/*.proto"},
+	}
+
+	// matches validate.proto which is needed by envoy protos
+	ValidateProtoMatcher = &GoModImport{
+		Package:  "github.com/envoyproxy/protoc-gen-validate",
+		Patterns: []string{"validate/*.proto"},
+	}
+
+	// matches all solo-kit protos, useful for any projects using solo-kit
+	SoloKitProtoMatcher = &GoModImport{
+		Package:  "github.com/solo-io/solo-kit",
+		Patterns: []string{"api/**/*.proto", "api/" + SoloKitMatchPattern},
+	}
+
+	// matches gogo.proto, used for gogoproto code gen.
+	GogoProtoMatcher = &GoModImport{
+		Package:  "github.com/gogo/protobuf",
+		Patterns: []string{"gogoproto/*.proto"},
+	}
+
+	// default match options which should be used when creating a solo-kit project
+	DefaultMatchOptions = []*Import{
+		{
+			ImportType: &Import_GoMod{
+				GoMod: ExtProtoMatcher,
+			},
+		},
+		{
+			ImportType: &Import_GoMod{
+				GoMod: ValidateProtoMatcher,
+			},
+		},
+		{
+			ImportType: &Import_GoMod{
+				GoMod: SoloKitProtoMatcher,
+			},
+		},
+		{
+			ImportType: &Import_GoMod{
+				GoMod: GogoProtoMatcher,
+			},
+		},
+	}
+)
+
+type goModOptions struct {
+	MatchOptions  []*GoModImport
+	LocalMatchers []string
+}
+
+// struct which represents a go module package in the module package list
+type Module struct {
+	ImportPath     string
+	SourcePath     string
+	Version        string
+	SourceVersion  string
+	Dir            string   // full path, $GOPATH/pkg/mod/
```

Comment:
what is Dir? this PR would be a lot easier to review with comments for some of these new types / interfaces